WcaOnRails/app/models/competition.rb

```diff
@@ -496,75 +497,41 @@ def pending_results_or_report(days)
     self.end_date < (Date.today - days) && (self.delegate_report.posted_at.nil? || results_posted_at.nil?)
   end
 
-  def start_date
-    year == 0 || month == 0 || day == 0 ? nil : Date.new(year, month, day)
-  end
-
-  def end_date
-    endYear == 0 || endMonth == 0 || endDay == 0 ? nil : Date.new(endYear, endMonth, endDay)
-  end
-
   private def unpack_dates
-    if @start_date.nil? && !start_date.blank?
-      @start_date = start_date.strftime("%F")
-    end
-    if @start_date.blank?
-      self.year = self.month = self.day = 0
+    if start_date
+      self.year = start_date.year
+      self.month = start_date.month
+      self.day = start_date.day
     else
-      unless /\A\d{4}-\d{2}-\d{2}\z/.match(@start_date)
-        errors.add(:start_date, I18n.t('common.errors.invalid'))
-        return false
-      end
-      self.year, self.month, self.day = @start_date.split("-").map(&:to_i)
-      unless Date.valid_date? self.year, self.month, self.day
-        errors.add(:start_date, I18n.t('common.errors.invalid'))
-        return false
-      end
-    end
-    if @end_date.nil? && !end_date.blank?
-      @end_date = end_date.strftime("%F")
+      self.year = self.month = self.day = 0
     end
-    if @end_date.blank?
-      self.endYear = self.endMonth = self.endDay = 0
+
+    if end_date
+      self.endYear = end_date.year
+      self.endMonth = end_date.month
+      self.endDay = end_date.day
     else
-      unless /\A\d{4}-\d{2}-\d{2}\z/.match(@end_date)
-        errors.add(:end_date, I18n.t('common.errors.invalid'))
-        return false
-      end
-      self.endYear, self.endMonth, self.endDay = @end_date.split("-").map(&:to_i)
-      unless Date.valid_date? self.endYear, self.endMonth, self.endDay
-        errors.add(:end_date, I18n.t('common.errors.invalid'))
-        return false
-      end
+      self.endYear = self.endMonth = self.endDay = 0
     end
   end
 
   private def dates_must_be_valid
-    if !confirmed_or_visible? && [year, month, day, endYear, endMonth, endDay].all? { |n| n == 0 }
-      # If the user left both dates empty, that's a-okay.
+    if start_date.nil? && end_date.nil?
+      if confirmed_or_visible?
+        errors.add(:start_date, I18n.t('common.errors.invalid'))
```

Comment:
That wouldn't do the same thing.

This code is a little odd.  You can think of the `return` as an `else` to `if confirmed_or_visible?`.